File: R/transform.R
#' log2 transform with an adaptive per-feature pseudocount
#'
#' Applies `log2(x + c)` where the pseudocount `c` is chosen *per row*
#' (i.e. per feature) as half the smallest finite positive value in that
#' row whenever the row contains a zero, and is `0` (no offset) otherwise.
#' This is the same offset rule the package's GSEA / QC / DE layers use
#' internally before any statistical test (DESIGN.md §4.5, decisions.md
#' #25-26).
#'
#' Conventions:
#'   - `x` is treated row-wise: rows are features (metabolites), columns
#'     are samples. Pass `t(...)` if your input has features as columns.
#'   - A vector input is treated as one feature.
#'   - Inf / NaN cells pass through (log2 of Inf is Inf, of NaN is NaN).
#'   - If `pseudocount` is supplied explicitly, the same scalar is added
#'     to every cell (legacy "global" behaviour, retained for callers
#'     that want to override the per-feature default).
#'
#' @param x A numeric matrix (rows = features) or a numeric vector
#'   (one feature).
#' @param pseudocount Optional scalar. If given, used directly for every
#'   cell; otherwise the per-feature half-min rule applies.
#' @return A matrix or vector of the same shape as `x`, log2-transformed.
#' @examples
#' x <- rbind(low  = c(0, 0.1, 0.5, 1),
#'            high = c(100, 200, 0, 400))
#' log2_with_pseudocount(x)
#' @export
log2_with_pseudocount <- function(x, pseudocount = NULL) {
  if (!is.null(pseudocount)) {
    return(log2(x + pseudocount))
  }
  if (!is.matrix(x)) {
    return(.log2_one_row(x))
  }
  out <- x
  for (i in seq_len(nrow(x))) {
    out[i, ] <- .log2_one_row(x[i, ])
  }
  out
}

.log2_one_row <- function(row) {
  has_zero <- any(row == 0, na.rm = TRUE)
  if (!has_zero) return(log2(row))
  finite_pos <- row[is.finite(row) & row > 0]
  pc <- if (length(finite_pos) > 0L) min(finite_pos) / 2 else 0
  log2(row + pc)
}

File: python/metabosetr/transform.py
"""log2 transform with an adaptive per-feature pseudocount.

Port of ``transform.R`` (DESIGN.md §4.5). ``log2(x + c)`` where the
pseudocount ``c`` is chosen *per row* (per feature) as half the smallest
finite positive value in that row whenever the row contains a zero, and is
``0`` otherwise.

Conventions (identical to the R version):

* ``x`` is treated row-wise: rows are features, columns are samples. Pass
  ``x.T`` if your input has features as columns.
* A 1-D input is treated as one feature.
* Inf / NaN cells pass through.
* If ``pseudocount`` is given, that scalar is added to every cell.
"""

from __future__ import annotations

import numpy as np


def _log2_one_row(row: np.ndarray) -> np.ndarray:
    row = np.asarray(row, dtype=float)
    has_zero = np.any(row == 0)  # NaN == 0 is False, matching R na.rm behaviour
    with np.errstate(divide="ignore", invalid="ignore"):
        if not has_zero:
            return np.log2(row)
        finite_pos = row[np.isfinite(row) & (row > 0)]
        pc = finite_pos.min() / 2 if finite_pos.size > 0 else 0.0
        return np.log2(row + pc)


def log2_with_pseudocount(x, pseudocount=None):
    """log2-transform ``x`` with the per-feature half-min pseudocount rule.

    Returns an array the same shape as ``x``.
    """
    if pseudocount is not None:
        with np.errstate(divide="ignore", invalid="ignore"):
            return np.log2(np.asarray(x, dtype=float) + pseudocount)
    arr = np.asarray(x, dtype=float)
    if arr.ndim == 1:
        return _log2_one_row(arr)
    out = np.empty_like(arr, dtype=float)
    for i in range(arr.shape[0]):
        out[i, :] = _log2_one_row(arr[i, :])
    return out
